libr/util/ht.c

```diff
@@ -102,9 +98,10 @@ static const struct {
     { 2147483648ul,	2362232233ul,	2362232231ul}
 };
 
-#define entry_is_free(x) (!x || !x->data)
-#define entry_is_deleted(x) x->data==&deleted_data
-#define entry_is_present(x) (x->data && x->data != &deleted_data)
+#define DELETED_HASH ((ut32)~0)
+#define entry_is_free(x) (!x->hash && !x->data)
+#define entry_is_deleted(x) (x->hash == DELETED_HASH && !x->data)
```

Comment:
Awesome explanations :) thanks

So it looks good to me then :)

> On 01 Jul 2016, at 15:19, resilar <notifications@github.com> wrote:
> 
> In libr/util/ht.c:
> 
> > @@ -102,9 +98,10 @@ static const struct {
> >      { 2147483648ul,	2362232233ul,	2362232231ul}
> >  };
> >  
> > -#define entry_is_free(x) (!x || !x->data)
> > -#define entry_is_deleted(x) x->data==&deleted_data
> > -#define entry_is_present(x) (x->data && x->data != &deleted_data)
> > +#define DELETED_HASH ((ut32)~0)
> > +#define entry_is_free(x) (!x->hash && !x->data)
> > +#define entry_is_deleted(x) (x->hash == DELETED_HASH && !x->data)
> Actually, we can have ht->size == NULL in the ht_(insert) function if it triggers rehashing and ht_(rehash) fails on calloc:
> 
> 137     ht->table = calloc (hash_sizes[new_size_index].size, sizeof (*ht->table));
> 138     if (!ht->table)
> 139         return;
> However, the !x check would not catch this unless hash_address = hash % ht->size also happens to be zero. We should properly handle the case that calloc fails (e.g., abort()).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>